File: typescript/patterns-use-cases/src/sagas/activities/car_rental_service.ts
import * as restate from "@restatedev/restate-sdk";

export const carRentalService = restate.service({
    name: "CarRentalService",
    handlers: {
        reserve: async (ctx: restate.Context, _req: { pickupLocation: string, rentalDate: string }) => {
            const carBookingId = ctx.rand.uuidv4();
            console.info(`Car ${carBookingId} reserved`);
            return carBookingId;
        },

        confirm: async (ctx: restate.Context, req: {carBookingId: string}) => {
            console.info(`Car ${req.carBookingId} confirmed`);
        },

        cancel: async (ctx: restate.Context, req: {carBookingId: string}) => {
            console.info(`Car ${req.carBookingId} cancelled`);
        },
    }
});

export type CarRentalService = typeof carRentalService;

File: go/patterns-use-cases/src/sagas/activities/carrentalservice.go
package activities

import (
	restate "github.com/restatedev/sdk-go"
	"log/slog"
)

type CarBookingRequest struct {
	PickupLocation string `json:"pickup_location"`
	RentalDate     string `json:"rental_date"`
}

type CarRentalService struct{}

func (CarRentalService) Reserve(ctx restate.Context, _carBookingRequest CarBookingRequest) (string, error) {
	carBookingId := restate.Rand(ctx).UUID().String()
	slog.Info("Car reserved: " + carBookingId)
	return carBookingId, nil
}

func (CarRentalService) Confirm(ctx restate.Context, carBookingId string) error {
	slog.Info("Car confirmed: " + carBookingId)
	return nil
}

func (CarRentalService) Cancel(ctx restate.Context, carBookingId string) error {
	slog.Info("Car cancelled: " + carBookingId)
	return nil
}
